lib/vrl/tests/tests/diagnostics/scoping_undefined_variable.vrl

```diff
@@ -0,0 +1,51 @@
+# result:
+#
+# error[E701]: call to undefined variable
+#    ┌─ :20:27
+#    │
+# 20 │     result = push(result, baz)
+#    │                           ^^^
+#    │                           │
+#    │                           undefined variable
+#    │                           did you mean "bar"?
+#    │
+#    = see language documentation at https://vrl.dev
+#
+# error[E701]: call to undefined variable
+#    ┌─ :25:23
+#    │
+# 25 │ result = push(result, bar)
+#    │                       ^^^
+#    │                       │
+#    │                       undefined variable
+#    │                       did you mean "foo"?
+#    │
+#    = see language documentation at https://vrl.dev
+
+foo = 5
+result = [foo]
+
+{
+  # overwrite existing variable
+  foo = 6
+
+  # instantiate new variable
+  bar = 7
+
+  # "mutate" existing variable
+  result = push(result, bar)
+
+  # nested blocks
+  {
+    bar = 8
+
+    # undefined variable
+    result = push(result, baz)
+  }
+}
+
+# variable only exists in inner scope
+result = push(result, bar)
+
+# variable is overwritten in inner scope
+push(result, foo)
```

Comment:
I don't quite follow what you're suggesting.

Reply:
I might have worded it confusingly. I meant asserting on the runtime values additionally to the types, e.g. that `foo` is `6` at the end of this (so that we manifest scoping behavior without let bindings and that arrays can be modified out of scope)

Reply:
Ah, right. Isn't that what we're testing in `lib/vrl/tests/tests/expressions/block/scoping.vrl`, though?

Reply:
Oh.. yeah :D I think I got completely blinded by the fact that the first line is a comment and mentally skipped it.

Is there a way to assert on local (both script scope and block scope) variables as well?

Reply:
That's what we validated in this test. We try to push `bar` on line 48, which results in an "undefined variable" error because that variable is local to the block scope. I'm unsure what you mean with "script scope", the top-level scope is always accessible/mutable in all nested scopes.

Reply:
> the top-level scope is always accessible/mutable in all nested scopes

Makes sense, I got pretty confused here. Thanks for clearing up 👍

Reply:
Reading this test and the `scoping.vrl` thing in detail (I should have done that earlier 😆) I see that you definitely covered all cases I would be interested in. At first I missed that the `push` call also resolved array it is pushing to. Very neat!